Fix NPEs due to disabled source

qwerty4030 opened the pull request:
This commit fixes some NPEs caused by implicitly performing a get request for a document that exists with its _source disabled and then trying to access the source. Instead of causing an NPE the following queries will throw an exception with a "source disabled" message (similar behavior as if the document does not exist).:
- GeoShape query for pre-indexed shape (throws IllegalArgumentException)
- Percolate query for an existing document (throws IllegalArgumentException)

A Terms query with a lookup will ignore the document if the source does not exist (same as if the document does not exist).

The GET and HEAD requests to [fetch the _source directly](https://www.elastic.co/guide/en/elasticsearch/reference/master/docs-get.html#_source) will return a 404 if the source is disabled (even if the document exists).

Updated tests (all passing) in  :core and :modules:percolator.
Added REST test `rest-api-spec/test/get_source/85_source_missing`.

Note: this bug still exists in the old Percolate and Multi Percolate APIs, but IMO not work fixing/testing since they are deprecated.

qwerty4030 commented:
To reproduce index a doc with _source disabled:
```
PUT test-index
{
  "mappings": {
    "test-type": {
      "_source": {
        "enabled": false
      }
    }
  }
}

PUT test-index/test-type/1
{}
```
Then try any of the following requests that implicitly run a get request and try to access the source: 
```
POST test-index/_search
{
  "query": {
    "terms": {
      "field": {
        "index": "test-index",
        "type": "test-type",
        "id": "1",
        "path": "test-field"
      }
    }
  }
}

POST test-index/_search
{
  "query": {
    "geo_shape": {
      "location": {
        "indexed_shape": {
          "index": "test-index",
          "type": "test-type",
          "id": "1",
          "path": "test-field"
        }
      }
    }
  }
}

POST test-index/_search
{
  "query": {
    "percolate": {
      "field": "query",
      "document_type": "test-type",
      "index": "test-index",
      "type": "test-type",
      "id": "1"
    }
  }
}

GET test-index/test-type/1/_source
```
Reponse is an NPE:
```
{
   "error": {
      "root_cause": [
         {
            "type": "null_pointer_exception",
            "reason": null
         }
      ],
      "type": "search_phase_execution_exception",
      "reason": "all shards failed",
      "phase": "query",
      "grouped": true,
      "failed_shards": [
         {
            "shard": 0,
            "index": "test-index",
            "node": "GImAaTDTTTW5h0lRacp-ng",
            "reason": {
               "type": "null_pointer_exception",
               "reason": null
            }
         }
      ],
      "caused_by": {
         "type": "null_pointer_exception",
         "reason": null
      }
   },
   "status": 500
}
```
Also the following request returns a 200, when it should be a 404:
```
HEAD test-index/test-type/1/_source
```
This PR updates the code to throw an exception with a more useful message (percolate and geo):
```
{
   "error": {
      "root_cause": [
         {
            "type": "illegal_argument_exception",
            "reason": "Shape with ID [1] in type [test-type] source disabled"
         }
      ],
      "type": "search_phase_execution_exception",
      "reason": "all shards failed",
      "phase": "query",
      "grouped": true,
      "failed_shards": [
         {
            "shard": 0,
            "index": "test-index",
            "node": "HKM5UtbeQlyE2lULZvn4Mg",
            "reason": {
               "type": "illegal_argument_exception",
               "reason": "Shape with ID [1] in type [test-type] source disabled"
            }
         }
      ],
      "caused_by": {
         "type": "illegal_argument_exception",
         "reason": "Shape with ID [1] in type [test-type] source disabled"
      }
   },
   "status": 400
}
```
In case someone was wondering how I found this: I was working on #18563 and used an old version of the [create index docs](https://www.elastic.co/guide/en/elasticsearch/reference/2.1/indices-create-index.html)(since fixed) that had an example with `_source` disabled in the mapping. I ran into this NPE when trying to use the old percolate API on an indexed doc. I found and fixed all places in the code that tried to access the _source from a get request without checking if it was null/empty.

clintongormley commented:
@martijnvg could you review this please?

martijnvg commented on modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java:
```diff
@@ -359,6 +359,11 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) throws I
             );
         }
+        if(getResponse.isSourceEmpty()) {
+            throw new ElasticsearchException(
```
maybe throw a `IllegalArgumentException` like in `GeoShapeQueryBuilder.java`?

qwerty4030 replied:
@martijnvg Sounds good. What about if the doc does not exist? Currently it throws a `ResourceNotFoundException` which means the search response status code is a 404. Not ideal since a 404 status also means the searched index does not exist. If its OK with you I'll change it throw a `IllegalArgumentException` as well. This will match the behavior in `GeoShapeQueryBuilder` so if the doc and/or `_source` is missing an `IllegalArgumentException` will be thrown (400 status code).

martijnvg replied:
Technically the _source doesn't exist, so a 404 could be returned. However the reason it is missing is because of source being disabled and then the 'percolate existing document' feature just can't work. For that reason I think 400 is a better response code, because that tells the client, that it is doing something that just doesn't work. So lets throw a `IllegalArgumentException ` here.

If a the document to be percolated is just missing, that document can be added, but enabling _source isn't possible on an existing index. So I think a 404 code is appropriate if the document to be percolated is missing.

martijnvg commented on core/src/main/java/org/elasticsearch/rest/action/get/RestHeadSourceAction.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Identical to {@link RestHeadAction} except this handler checks if the document and its _source exist.
+ */
+public class RestHeadSourceAction extends BaseRestHandler {
```
maybe instead of adding this new rest action the existing `ResHeadAction.java` should be modified?

I think we do something like this in `ResHeadAction.java`:

```
ActionListener listener;
if (request.path().endsWith("_source")) {
listener = new RestResponseListener<GetResponse>(channel) {
            @Override
            public RestResponse buildResponse(GetResponse response) {
                if (!response. isSourceEmpty()) {
                    return new BytesRestResponse(NOT_FOUND, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY);
                } else {
                    return new BytesRestResponse(OK, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY);
                }
            }
        }
} else {
listener = new RestResponseListener<GetResponse>(channel) {
            @Override
            public RestResponse buildResponse(GetResponse response) {
                if (!response.isExists()) {
                    return new BytesRestResponse(NOT_FOUND, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY);
                } else {
                    return new BytesRestResponse(OK, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY);
                }
            }
        };
}
```

martijnvg replied:
A rest test for this would be great too (under `rest-api-spec/test/get_source/`), but I can add that too.

qwerty4030 replied:
@martijnvg I will add some rest tests, or at least try :) Are they run as part of `gradle core:test`?
I created `RestHeadSourceAction` for a couple reasons: The GET actions were already split into 2 files: `RestGetAction` and `RestGetSourceAction` (IMO not ideal due to duplicated code, but I copied that pattern). Also `request.path().endsWith("_source")` is `false` for `HEAD test-index/test-type/1/_source/` (and also for `HEAD test-index/test-type/1/_source//` which may be a bug?) Is it possible to use the code in [PathTrie.java#L212](https://github.com/elastic/elasticsearch/blob/78d615f32054baeede280263ebc76bd11ea3951d/core/src/main/java/org/elasticsearch/common/path/PathTrie.java#L212) to break up the path by the separator and get the last part? If not I can write that logic myself and reuse `RestHeadAction` for `_source` as well. Would something like this work?
```
final boolean source = request.path().endsWith("_source"); // TODO figure out how to handle /_source/
client.get(getRequest, new RestResponseListener<GetResponse>(channel) {
    @Override
    public RestResponse buildResponse(GetResponse response) {
        if (!response.isExists()) {
            return new BytesRestResponse(NOT_FOUND, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY);
        } else if (source && response.isSourceEmpty()) { // doc exists, but source might not
            return new BytesRestResponse(NOT_FOUND, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY);
        } else {
            return new BytesRestResponse(OK, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY);
        }
    }
});
```
I also had another idea to do something clever in the `controller.registerHandler` call to distinguish the HEAD doc request vs HEAD doc _source request.

martijnvg replied:
> Are they run as part of gradle core:test?

No, the tests run as part of: `gradle :distribution:integ-test:integTest`. However I now realize that we don't have any rest integration test for HEAD actions... this doesn't seem to work at the moment too. So adding rest tests for head actions should be done another time.

> The GET actions were already split into 2 files: RestGetAction and RestGetSourceAction

True, these have different files, but the there also more difference between the two. The get request supports `fields` option while get source does not.

> Also request.path().endsWith("_source") is false for HEAD test-index/test-type/1/_source/ (and also for HEAD test-index/test-type/1/_source// which may be a bug?) 

Agreed, that does lead to bugs.

> I also had another idea to do something clever in the controller.registerHandler call to distinguish the HEAD doc request vs HEAD doc _source request. Maybe something like this?

I like that approach. However the `RestHeadAction.java` does need to extend from `RestHandler` otherwise it can't registered in `NetworkModule.java`. Maybe we should do something like [this](https://gist.github.com/martijnvg/6a28e57e6b39be2cdf4044a3dbc85669)? We would then also need to modify the  `NetworkModule.java` file to register the concrete subclasses instead.

martijnvg commented:
Thanks @qwerty4030, this looks good! I left a couple of comments.

martijnvg commented:
@qwerty4030 Your update looks good! Can you also resolve the merge conflicts? I'll will then happily merge it .

qwerty4030 commented:
@martijnvg awesome! Thanks for the review. Do you suggest merging master into my branch or rebase my branch on master?

martijnvg commented:
Since the change is small, lets rebase your branch to master?

qwerty4030 commented:
@martijnvg rebased and squashed the commit. Can you take a look a this REST test I added: `rest-api-spec/test/get_source/85_source_missing` Thanks!
All the following tests pass for me:

- `gradle :distribution:integ-test-zip:integTest`
- `gradle :core:test`
- `gradle :core:integTest`
- `gradle :modules:perolcator:test`
- `gradle :modules:perolcator:integTest`

qwerty4030 commented:
Oh and thanks for the tip about registering the handlers in `NetworkModule`. Totally missed that in the initial commit! What exactly is the purpose of that module? When I forgot to register the new handler class everything seemed to work OK? I poked around the code and noticed `RestIndexAction.CreateHandler` is not registered in `NetworkModule` (howerver `RestIndexAction` is).

martijnvg commented:
> Can you take a look a this REST test I added: rest-api-spec/test/get_source/85_source_missing

This test looks good. Thanks you!

> What exactly is the purpose of that module? 

It instantiates the rest actions, netty/local transport and their infrastructure.

> When I forgot to register the new handler class everything seemed to work OK?

I think that because in the previous version of this PR the RestHeadAction itself was registered in the network module and in the constructor (like RestIndexAction) additional handles were registered.
